userspace/async/async_metadata_source.tpp

```diff
@@ -0,0 +1,286 @@
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#include "logger.h"
+
+#include <assert.h>
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <list>
+#include <string>
+
+namespace sysdig
+{
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::async_metadata_source(
+		const uint64_t max_wait_ms,
+		const uint64_t ttl_ms):
+	m_max_wait_ms(max_wait_ms),
+	m_ttl_ms(ttl_ms),
+	m_thread(),
+	m_running(false),
+	m_terminate(false),
+	m_mutex(),
+	m_queue_not_empty_condition(),
+	m_metadata_map()
+{ }
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::~async_metadata_source()
+{
+	try
+	{
+		stop();
+	}
+	catch(...)
+	{
+		g_logger.log(std::string(__FUNCTION__) +
+		             ": Exception in destructor",
+		             sinsp_logger::SEV_ERROR);
+	}
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_max_wait() const
+{
+	return m_max_wait_ms;
+}
+
+template<typename key_type, typename metadata_type>
+uint64_t async_metadata_source<key_type, metadata_type>::get_ttl() const
+{
+	return m_ttl_ms;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::stop()
+{
+	bool join_needed = false;
+
+	{
+		std::unique_lock<std::mutex> guard(m_mutex);
+
+		if(m_running)
+		{
+			m_terminate = true;
+			join_needed = true;
+
+			// The async thread might be waiting for new events
+			// so wake it up
+			m_queue_not_empty_condition.notify_one();
+		}
+	} // Drop the mutex before join()
+
+	if (join_needed)
+	{
+		m_thread.join();
+
+		// Remove any pointers from the thread to this object
+		// (just to be safe)
+		m_thread = std::thread();
+	}
+}
+
+template<typename key_type, typename metadata_type>
+bool async_metadata_source<key_type, metadata_type>::is_running() const
+{
+	std::lock_guard<std::mutex> guard(m_mutex);
+
+	return m_running;
+}
+
+template<typename key_type, typename metadata_type>
+void async_metadata_source<key_type, metadata_type>::run()
+{
+	m_running = true;
+
+	while(!m_terminate)
+	{
+		{
+			std::unique_lock<std::mutex> guard(m_mutex);
+
+			while(!m_terminate && m_request_queue.empty())
+			{
+				// Wait for something to show up on the queue
+				m_queue_not_empty_condition.wait(guard);
+			}
+
+			prune_stale_requests();
+		}
+
+		if(!m_terminate)
+		{
+			run_impl();
```

Comment:
you've released the lock before invoking this function. i see that you reacquire when doing queue operations, but it makes this particular function appear suspicious. We do all the checks against the CV in a way that makes it seem that tight synchronization is essential, then we release the lock before making this call.

If we don't actually need it, I think we need to simplify the above half of the function
- don't bother with the double check locking. termination is something we expect to happen once per ever
- don't bother checking for the empty queue around the CV wait

So pretty much lock->check for terminate->CV wait is all we need if we're not holding the lock across the call to run_impl